### PR TITLE
MSVC_part2

### PR DESCRIPTION
**Pull request by dkeeney**
unit tests run under MSVC now;  Including HelloRegionTest and CppRegionTest.
This push is to make sure it did not break Ubuntu or OSx.

### CODE REVIEW

**Comment by dkeeney** [comment]
I noticed that many of the unit tests contain something about a Manual verification.  This will not happen.  These tests must be fully automated.  Find a way to programmatically verify that the output data is correct.

**Comment by dkeeney** [comment]
> error: Microsoft Visual C++ 9.0 is required. Get it from http://aka.ms/vcpython27

It looks like we will not be able to build the .py bindings for Python2.7 when running under MSVC. We would have to downgrade to MSVC 9.0 in order to create a compatible library and MSVC 9.0 will not compile C++11.  I cannot find any work-around.

So it looks like it is Python3.x only for MSVC.

**Comment by ctrl-z-9000-times** [comment]
>  I cannot find any work-around.

That's fine. It works on a range of versions so there is at least one version which works on every platform.

**Comment by dkeeney** [comment]
I can now locally compile and run .py unit tests using MSVC and Python 3.6.  Now I just need to get it to run that way in Appveyor CI.

**Comment by dkeeney** [comment]
Switching to Python 3.6.6 seemed to work better.  I can install numpy now.

But the unit tests run slower under MSVC in the appveyor environment.
```
[----------] 1 test from HelloSPTPTest
[ RUN      ] HelloSPTPTest.performance
starting test. DIM_INPUT=10000, DIM=2048, CELLS=10
EPOCHS = 5000
starting:  5000 iterations.Epoch = 4999
Anomaly = 0.488755
SP= 1418,1421,1422,1424,1426,1429,1431,1432,1433,1434
TP= 1418,1421,1422,1424,1426,1429,1431,1432,1433,1434
==============TIMERS============
Init:	0.10218
Random:	0.001916
Encode:	0.0161255
SP:	1.00381
TP:	2.70258
AN:	9.33096
Total elapsed time = 13 seconds
ERR:  CHECK FAILED: "timeTotal <= CI_avg_time" HelloSPTP test slower than expected! (13,should be 9 [c:\projects\nupic-cpp\src\examples\hellosptp.cpp line 143]
unknown file: error: C++ exception with description "CHECK FAILED: "timeTotal <= CI_avg_time" HelloSPTP test slower than expected! (13,should be 9" thrown in the test body.
[  FAILED  ] HelloSPTPTest.performance (13188 ms)
[----------] 1 test from HelloSPTPTest (13188 ms total)
```

**Comment by dkeeney** [comment]
OK, MSVC build for Python3.6 officially works. 😎

- There are still some warnings in the bindings compile (mostly type or sign mismatch)
- Still have a lot of .py tests that are skipped.
- Serialization that crosses the C++/Py boundary needs some work.

These can be addressed in separate PRs.
When you have time, please review this PR and lets get it checked in.

**Comment by dkeeney** [comment]
@breznak I assume you have access to the UI for the CI ... can you see if there is something that is causing two instances of appveyor to run every time we push.  They appear to be the same and we only need one.

File: external/YamlCppLib.cmake [hunk 1]
```diff
@@ -24,14 +24,17 @@
 if(EXISTS ${REPOSITORY_DIR}/build/ThirdParty/share/yaml-cpp-0.6.2.tar.gz)
     set(URL ${REPOSITORY_DIR}/build/ThirdParty/share/yaml-cpp-0.6.2.tar.gz)
 else()
-    set(URL https://github.com/jbeder/yaml-cpp/archive/yaml-cpp-0.6.2.tar.gz)
+    #set(URL https://github.com/jbeder/yaml-cpp/archive/yaml-cpp-0.6.2.tar.gz)
+    set(REPO https://github.com/jbeder/yaml-cpp.git)
```
Comment (breznak) [inline external/YamlCppLib.cmake]: is this needed, safe to use master for dependencies?

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -342,10 +342,6 @@ install(TARGETS
 install(FILES ${PROJECT_BINARY_DIR}/Version.hpp
         DESTINATION include/nupic)
 
-# install Data.csv for a test
-install(FILES "${PROJECT_SOURCE_DIR}/test/unit/engine/Data.csv"
-        DESTINATION "data/")
```
Comment (breznak) [inline src/CMakeLists.txt]: where is this then copied?

File: src/examples/HelloSPTP.cpp [hunk 3]
```diff
@@ -139,7 +139,11 @@ void run() {
       const size_t timeTotal = (size_t)floor(tAll.getElapsed());
       cout << "Total elapsed time = " << timeTotal << " seconds" << endl;
       #ifdef NDEBUG
-        const size_t CI_avg_time = (size_t)floor(7*Timer::getSpeed()); //sec
+	    #ifdef _MSC_VER
+          const size_t CI_avg_time = (size_t)floor(14*Timer::getSpeed()); //sec
+		#else
+          const size_t CI_avg_time = (size_t)floor(7*Timer::getSpeed()); //sec
```
Comment (breznak) [inline src/examples/HelloSPTP.cpp]: 👍 nice that you made 2 versions of the check, still keeping the unix test useful

File: src/test/unit/engine/CppRegionTest.cpp [hunk 4]
```diff
@@ -380,27 +378,23 @@ Network helperRealmain() {
...
-            << std::endl;
+  EXPECT_DOUBLE_EQ(rval, 30.1) << " after setting to 30.1";
 
   // --- test getParameterInt64Array ---
   // Array a is not allocated by us. Will be allocated inside getParameter
   Array a(NTA_BasicType_Int64);
   level1->getParameterArray("int64ArrayParam", a);
-  std::cout << "level1.int64ArrayParam size = " << a.getCount() << std::endl;
-  std::cout << "level1.int64ArrayParam = [ ";
+  EXPECT_EQ(a.getCount(), 4u) << "size set in initializer of TestNode is 4";
```
Comment (breznak) [inline src/test/unit/engine/CppRegionTest.cpp]: 👍 for reducing the tests' output add adding a check with failover-message.

File: src/test/unit/engine/HelloRegionTest.cpp [hunk 5]
```diff
@@ -36,57 +38,68 @@ namespace testing {
...
-//    EXPECT_FLOAT_EQ(buffer[0], 0); //TODO add test, which values should be here?
-    std::cout << "  " << i << "    " << buffer[i] << "" << std::endl;
-  }
-
+  ArrayRef outputArray = region->getOutputData("dataOut");
+  EXPECT_TRUE(outputArray.getType() == NTA_BasicType_Real32);
+  EXPECT_EQ(outputArray.getCount(), testdata[0].size());
+  const Real32 *buffer = (const Real32 *)outputArray.getBuffer();
+  for (size_t i = 0; i < outputArray.getCount(); i++)
+    EXPECT_FLOAT_EQ(buffer[i], testdata[0][i]);
```
Comment (breznak) [inline src/test/unit/engine/HelloRegionTest.cpp]: great to see this finally fixed. The Data.csv and "weird" values in this test have been a pain

File: src/test/unit/engine/HelloRegionTest.cpp [hunk 6]
```diff
@@ -98,21 +111,28 @@ TEST(HelloRegionTest, demo) {
...
-  std::shared_ptr<Region> region2 = net2.getRegion("region"); //TODO add more checks and asserts here
-  region2->executeCommand(loadFileArgs);
+  std::shared_ptr<Region> region2 = net2.getRegion("region");
   ArrayRef outputArray2 = region2->getOutputData("dataOut");
-  const Real64 *buffer2 = (const Real64 *)outputArray2.getBuffer();
 
-  net.run(DATA_SIZE);
-  net2.run(DATA_SIZE);
+  // fetch the data rows for both networks.
+  net.run((int)data_rows-1);
```
Comment (breznak) [inline src/test/unit/engine/HelloRegionTest.cpp]: why rows-1?

File: src/CMakeLists.txt (hunk 2 above)
Comment (breznak) [reply src/CMakeLists.txt]: ...the test generates the test data itself 👍

**Comment by breznak** [comment]
This resolves #69

File: external/YamlCppLib.cmake (hunk 1 above)
Comment (dkeeney) [reply external/YamlCppLib.cmake]: > is this needed, safe to use master for dependencies?

This is temporary.  By using the latest code we get the fix that prevents yaml-cpp from crashing in debug mode.  Also, I was looking carefully at the 0.6.2 download and saw that the changes listed in the release notes were not actually in the code. So I don't know what that file is but it is probably not the real 0.6.2 release.

As soon as they do another release I will switch back.

File: src/test/unit/engine/HelloRegionTest.cpp (hunk 5 above)
Comment (dkeeney) [reply src/test/unit/engine/HelloRegionTest.cpp]: >  Data.csv

By not having a prepared file we don't have to run the tests from that directory.  So no requirement to CD to build/Release/bin.

File: src/test/unit/engine/HelloRegionTest.cpp (hunk 6 above)
Comment (dkeeney) [reply src/test/unit/engine/HelloRegionTest.cpp]: > why rows-1?

Because the first row was already consumed before the save.

**Comment by dkeeney** [comment]
Thanks guys :)

I will be out the rest of today but back to work on Sat.
- I want to look at cleaning up the Skipped .py tests.
- I also need to integrate the new SDR class into the Array class so it can be passed between regions.

File: external/YamlCppLib.cmake (hunk 1 above)
Comment (breznak) [reply external/YamlCppLib.cmake]: >  By using the latest code we get the fix that prevents yaml-cpp from crashing in debug mode 

Great, just please open an Issue to turn it back to release later. 

>  Also, I was looking carefully at the 0.6.2 download and saw that the changes listed in the release notes were not actually in the code. So I don't know what that file is but it is probably not the real 0.6.2 release. 

Wow! 😮 Would you like to open a bug for their repo?

**Pull request merged**